### PR TITLE
Bluetooth: host: Use uintptr in bt_le_ext_adv_get_index before return

### PR DESCRIPTION
**Pull request by joerchan**
Align bt_le_per_adv_sync_get_index, bt_conn_index and
bt_le_ext_adv_get_index in use of uintptr_t.

This fixes an issue where the cast to uint8_t happened before the
assert for a valid index, which could lead to invalid pointers passing
this assert.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

### CODE REVIEW

**Comment by jhedberg** [comment]
Why is it important to use an integer type that can fit a pointer? The value itself is an index into an array so mixing something about pointers into the type is confusing, IMO. E.g. net_buf has ` int net_buf_id(struct net_buf *buf); ` which seems good enough to me.

**Comment by joerchan** [comment]
This is a long overdue follow-up. As Emil explained in his periodic advertising PR the first cast to uint8_t could actually end up passing the ASSERT even though the pointer is invalid. Since the subtract operation is done on pointer values, we should use an uintptr_t to check the result of this, before we return the value as an uint8_t.

Original discussion here:
https://github.com/zephyrproject-rtos/zephyr/pull/25326#discussion_r446541714

**Comment by jhedberg** [comment]
@joerchan so you mean net_buf API needs fixing as well?

**Comment by jhedberg** [comment]
@joerchan would it not make sense to have an ASSERT *before* doing the pointer subtraction, i.e. first check that the given pointer is in the right range? That should allow using whatever integer type is appropriate for the range of possible indices, right?

**Comment by joerchan** [comment]
@jhedberg As far as I can see net_buf_id does not have the problem, since the value is not checked.

Checking the input parameter for valid range would be the same result, we are just checking the range a bit different.

Maybe I don't understand what you mean? Is there a problem with the use of the uintptr_t for the intermediate variable? We are returning a uint8_t, so everything should be as before, we just don't downcast to uint8_t before the assert anymore. 

I see that bt_conn_index should be updated in a similar way, so I'll do that before we merge this PR.

**Comment by jhedberg** [comment]
@joerchan sorry, it's just me not reading the patch correctly. I didn't realise that the return value type isn't changed.

**Review by jhedberg — APPROVED** [review]

**Comment by joerchan** [comment]
@jhedberg Ok, my commit message could have been clearer so I have updated it. And made the same change to `bt_conn_index`

File: subsys/bluetooth/host/hci_core.c [hunk 1]
```diff
@@ -382,7 +382,7 @@ const bt_addr_le_t *bt_lookup_id_addr(uint8_t id, const bt_addr_le_t *addr)
 #if defined(CONFIG_BT_EXT_ADV)
 uint8_t bt_le_ext_adv_get_index(struct bt_le_ext_adv *adv)
 {
-	uint8_t index = adv - adv_pool;
+	uintptr_t index = adv - adv_pool;
```
Comment (Thalley) [inline subsys/bluetooth/host/hci_core.c:385]: Can we please add an explicit cast to `uint8_t` in the return statements? Just for good measure :)

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/hci_core.c (hunk 1 above)
Comment (Thalley) [reply subsys/bluetooth/host/hci_core.c:385]: This is apparently also missing in `bt_le_per_adv_sync_get_index`

Comment (joerchan) [reply subsys/bluetooth/host/hci_core.c:385]: Yes, added now.

**Review by Thalley — APPROVED** [review]
LGTM

File: subsys/bluetooth/host/conn.c [hunk 2]
```diff
@@ -2685,7 +2685,7 @@ int bt_conn_auth_pairing_confirm(struct bt_conn *conn)
 
 uint8_t bt_conn_index(struct bt_conn *conn)
 {
-	uint8_t index;
+	uintptr_t index;
```
Comment (pabigot) [inline subsys/bluetooth/host/conn.c:2688]: The strictly correct type for this would be `ptrdiff_t` as it's used for the difference between pointers (and is signed).

It appears that the result of subtraction is always non-negative, so the corresponding unsigned type is `size_t`.

`uintptr_t` is usually used for representing pointer values, not pointer differences.

**Review by pabigot — CHANGES_REQUESTED** [review]
This should use `size_t`.

File: subsys/bluetooth/host/conn.c (hunk 2 above)
Comment (Thalley) [reply subsys/bluetooth/host/conn.c:2688]: That's a very good point. 

Since we are subtracting two values, without checking if the value is going to be positive or negative, I guess the correct solution would require a signed variable. Alternative is to also add an __ASSERT if the pointer checked is lower than the pool that it's checked against.

Comment (pabigot) [reply subsys/bluetooth/host/conn.c:2688]: We generally don't bother sign-checking this sort of thing, since it's "known" that the pointer is a member of the sequence starting at the base pointer that's subtracted.  For MISRA reasons it should be `index = (size_t)(ptr - base);` but that's not generally done.

Comment (Thalley) [reply subsys/bluetooth/host/conn.c:2688]: > We generally don't bother sign-checking this sort of thing, since it's "known" that the pointer is a member of the sequence starting at the base pointer that's subtracted. 

If we know that the pointer is a member of the sequence, why bother checking the index at all? I thought the point of the check was to check if the pointer indeed was a member of the sequence. 

What value does the `size_t` typecast add here?

Comment (pabigot) [reply subsys/bluetooth/host/conn.c:2688]: OK, I was again misled because the deltas displayed don't cover enough of what the code does, so I didn't notice the asserts.

What problem are you trying to solve here?  The problem description is not detailed.

Presumably at some point somebody passed in a `conn` that was either **not** a member of the corresponding `foo_conns` array.

Was it a null pointer?  Was it a pointer to a valid `conn` but the `conn->type` incorrectly identified the sequence it was to belong to?  Or is it that somebody set `CONFIG_BT_MAX_FOO_CONN` greater than 127 so the truncation discarded important data?

Clearly `conn` can't be null, because then the switch argument would cause a fault.

I have to guess that what happened was either the `type` was wrong or there were more than 127 connections of the given type.  In any case, you need to check that the pointer is strictly within the range of the connection array, which means you have to check for negative values.

In short, if you want to do this correctly, you should do:

```
    ptrdiff_t index;
    ...
    index = conn - foo_conns;
    __ASSERT(index >=0 && index < CONFIG_BT_MAX_FOO_CONNS);
    ...
    return (uint8_t)index;
    ```
```

If you want to obscurely optimize away the check for negative values instead of letting the compiler do that, then do:

```

    size_t index;
    ...
    index = (size_t)(conn - foo_conns); // negatives become large positives
    __ASSERT(index < CONFIG_BT_MAX_FOO_CONNS);
    ...
    return (uint8_t)index;
    ```
```

The cast is required by MISRA because without it you'd be assigning a signed potentially negative value into a variable that can only hold an unsigned value, and you should be explicit that the value conversion is intentional.

Comment (Thalley) [reply subsys/bluetooth/host/conn.c:2688]: I think @pabigot proposed solution would indeed be the "correct" way of doing this. 

```
ptrdiff_t index;
...
index = conn - foo_conns;
__ASSERT(index >= 0 && index < CONFIG_BT_MAX_FOO_CONNS);
...
return (uint8_t)index;
```

Comment (joerchan) [reply subsys/bluetooth/host/conn.c]: @pabigot I was trying to fix the assert, which had the problem that it truncated before the assert, which could lead to passing the assert with a pointer that did not point to the foo_conns array.

The purpose of the assert is to check that the pointer is part of the sequence. I have updated to explicit check, and will let the compiler do the optimization instead of the obscure optimization.

Thank you for the detailed feedback.

**Review by pabigot — APPROVED** [review]
